bcrypt.js

```diff
@@ -82,6 +82,65 @@ module.exports.genSalt = function genSalt(rounds, minor, cb) {
     });
 };
 
+module.exports.genSaltByTime = function genSaltByTime(exptime, minor, cb) {
+    var error;
+    var rounds = 10;
+    // if callback is first argument, then use defaults for others
+    if (typeof arguments[0] === 'function') {
+        // have to set callback first otherwise arguments are overriden
+        cb = arguments[0];
+        exptime = 100;
+        minor = 'b';
+        // callback is second argument
+    } else if (typeof arguments[1] === 'function') {
+        // have to set callback first otherwise arguments are overriden
+        cb = arguments[1];
+        minor = 'b';
+    }
+
+    if (!cb) {
+        return promises.promise(genSaltByTime, this, [exptime, minor]);
+    }
+
+    // default 100 milliseconds and minimum 4 miliseconds
+    if (!exptime) {
+        exptime = 100;
+    } else if (exptime < 4) {
+        exptime = 4;
+    } else if (typeof exptime !== 'number') {
+        // callback error asynchronously
+        error = new Error('Expected time must be a number');
+        return process.nextTick(function() {
+            cb(error);
+        });
+    }
+
+    if (!minor) {
+        minor = 'b'
+    } else if (minor !== 'b' && minor !== 'a') {
+        error = new Error('minor must be either "a" or "b"');
+        return process.nextTick(function() {
+            cb(error);
+        });
+    }
+
+    crypto.randomBytes(16, function(error, randomBytes) {
+        if (error) {
+            cb(error);
+            return;
+        }
+
+        //since the relation b/w expected time and rounds roughly follows exptime = 2^(rounds-3)
```

Comment:
Source of this?

Reply:
While working with the code, I observed that the time taken for each round was going exponential. When I analyzed the actual time taken with the help of following snippet.
```(javascript)
var bcrypt=require('./bcrypt.js');
async function pass(){
    for (let rounds = 1; rounds <= 20; rounds++){
        var ini = new Date();
        var hashpass = await bcrypt.hashSync("Blueisthecolorofsky", rounds);
        var fin = new Date();
        console.log(rounds,fin-ini);
    }
}
pass();
```
and plotted the following graph with the help of Excel,

![Graph1](https://i.ibb.co/TbFN3sj/Time-in-ms-vs-No-of-rounds-2.png)   

and compared it with the graph of 2^(rounds) with different proportionality constants k:

![Graph2](https://i.ibb.co/zJPWtxM/Time-taken-in-ms.png)

where it was observed to be very close with constant k=1/8, leading us to believe that time taken for each round closely follows the formula:
```
expected_time = (1/8) * 2^(number_of_rounds)
```
Thus it was concluded that:  
```
number_of_rounds = log2(expected_time) + 3
```

Reply:
As processors go faster and architectures change, the `k` value is going to change, the best tool is to benchmark on your server

Reply:
Don't know much about this, but couldn't a max time be set then it will do more rounds until time capped--with a minimum value? or is bcrypt not built like this